File: yangxj96-serve/yangxj96-serve-auth/src/main/java/io/github/yangxj96/server/auth/service/impl/UserDetailsServiceImpl.java
package io.github.yangxj96.server.auth.service.impl;

import com.baomidou.mybatisplus.core.conditions.query.LambdaQueryWrapper;
import io.github.yangxj96.bean.user.User;
import io.github.yangxj96.server.auth.service.UserService;
import jakarta.annotation.Resource;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;

import java.util.Collections;


/**
 * 用户名密码登录
 *
 * @author 杨新杰
 * @date 2021/11/30 14:02
 */
@Slf4j
@Service
public class UserDetailsServiceImpl implements UserDetailsService {

    @Resource
    private UserService userService;

    @Override
    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
        if (StringUtils.isEmpty(username)) {
            throw new NullPointerException("用户名为空");
        }
        User user = userService.getOne(
                new LambdaQueryWrapper<User>()
                        .eq(User::getUsername, username)
                        .last("LIMIT 1")
        );
        if (null == user) {
            throw new UsernameNotFoundException("用户不存在");
        }
        // user.setAuthorities(userService.getAuthorities(user.getId()))
        user.setAuthorities(Collections.emptyList());
        return user;
    }

}

File: yangxj96-serve/yangxj96-serve-auth/src/main/java/io/github/yangxj96/server/auth/service/impl/UserDetailsServiceImpl.kt
package io.github.yangxj96.server.auth.service.impl

import com.baomidou.mybatisplus.core.conditions.query.QueryWrapper
import io.github.yangxj96.bean.user.User
import io.github.yangxj96.server.auth.service.UserService
import jakarta.annotation.Resource
import org.apache.commons.lang3.StringUtils
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.stereotype.Service

/**
 * 用户名密码登录的实现
 */
@Service
class UserDetailsServiceImpl : UserDetailsService {

    @Resource
    private lateinit var userService: UserService

    @Throws(UsernameNotFoundException::class)
    override fun loadUserByUsername(username: String): UserDetails {
        try {
            if (StringUtils.isEmpty(username)) {
                throw NullPointerException("用户名为空")
            }

            val wrapper = QueryWrapper<User>()
                .eq("username", username)
                .last("LIMIT 1")

            val user = userService.getOne(wrapper) ?: throw UsernameNotFoundException("用户不存在")

            user.authorities = userService.getAuthoritiesByUserId(user.id!!)

            return user
        } catch (e: Exception) {
            e.printStackTrace()
            throw e
        }
    }
}
